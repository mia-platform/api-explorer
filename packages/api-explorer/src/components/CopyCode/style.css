.mia-ctc-button span {
  text-transform: uppercase;
  background: transparent;
  color: #AAA;
  cursor: pointer;
  border: 0px;
}

.mia-ctc-button span:hover {
  color: white;
}
