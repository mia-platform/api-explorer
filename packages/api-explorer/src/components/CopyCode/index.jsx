import React from 'react'
import PropTypes from 'prop-types'
import {FormattedMessage} from 'react-intl'

import CopyText from '../CopyText'
import './style.css'


export default function CopyCode({code}) {
    return (
      <CopyText className="mia-ctc-button" text={code}>
        {({isCopied}) => (
          <span>
            {isCopied ?
              <span><FormattedMessage id="code.copied" defaultMessage="Copied" /></span> :
              <span><FormattedMessage id="code.copy" defaultMessage="Copy" /></span>
          }
          </span>
        )}
        
      </CopyText>
    )
}

CopyCode.propTypes = {
  code: PropTypes.string.isRequired,
};
