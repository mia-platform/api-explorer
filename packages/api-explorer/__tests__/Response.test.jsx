import {IntlProvider} from 'react-intl';
import CopyCode from '../src/components/CopyCode'

const React = require('react');
const { shallow, mount } = require('enzyme');
const petstore = require('./fixtures/petstore/oas');

const Response = require('../src/components/Response');
const ResponseMetadata = require('../src/components/Response/ResponseMetadata')
const Oas = require('../src/lib/Oas');

const { Operation } = Oas;
const oas = new Oas(petstore);

const props = {
  result: null,
  operation: new Operation({}, '/pet', 'post'),
  hideResults: () => {},
  oas,
  oauth: false,
};

describe('no result', () => {
  test('nothing should render', () => {
    const codeSampleResponseTabs = shallow(<Response {...props} />);

    expect(codeSampleResponseTabs.find('span').length).toBe(0);
  });
});

describe('setTab', () => {
  test('setTab should change state of selectedTab', () => {
    const doc = shallow(<Response {...props} />);

    expect(doc.state('responseTab')).toBe('result');

    doc.instance().setTab('metadata');

    expect(doc.state('responseTab')).toBe('metadata');

    doc.instance().setTab('result');

    expect(doc.state('responseTab')).toBe('result');
  });
});

describe('exampleTab', () => {
  test('exampleTab should change state of exampleTab', () => {
    const doc = shallow(<Response {...props} />);

    expect(doc.state('exampleTab')).toBe(0);

    doc.instance().setExampleTab(1);

    expect(doc.state('exampleTab')).toBe(1);
  });
});

test('and CopyCode is rendered correctly with a responseBody !== string', () => {
  const customProps = {
    ...props,
    result: {
      ...props.result,
      responseBody: {test: 1}
    }
  }
  const doc = shallow(<Response {...customProps} />);
  expect(doc.find(CopyCode).prop('code')).toEqual("{\"test\":1}")
})

test('and CopyCode is rendered correctly with a result binary', () => {
  const customProps = {
    ...props,
    result: {
      ...props.result,
      responseBody: {test: 1},
      type: 'application/json'
    }
  }
  const doc = shallow(<Response {...customProps} />);
  expect(doc.find(CopyCode).prop('code')).toEqual("{\"test\":1}")
  expect(doc.state().collapse).toEqual(undefined)

  const buttons = doc.find('.mia-ctc-button')
  const collapseButton = buttons.at(0)
  collapseButton.simulate('click', {preventDefault: jest.fn()})
  expect(doc.state().collapse).toEqual(true)

  const expandbutton = buttons.at(1)
  expandbutton.simulate('click', {preventDefault: jest.fn()})
  expect(doc.state().collapse).toEqual(false)
})

test('and CopyCode is rendered correctly with a responseBody === string', () => {
  const customProps = {
    ...props,
    result: {
      ...props.result,
      responseBody: 'testResponseBody'
    }
  }
  const doc = shallow(<Response {...customProps} />);
  expect(doc.find(CopyCode).prop('code')).toEqual(customProps.result.responseBody)
})

test('should show different component tabs based on state', () => {
  const wrapper = mount(
    <IntlProvider>
      <Response
        {...props}
        result={{
          status: 200,
          responseBody: JSON.stringify({ a: 1 }),
          requestBody: JSON.stringify({ b: 2 }),
          requestHeaders: [],
          method: 'post',
          responseHeaders: [],
        }}
      />
    </IntlProvider>
  );
  expect(wrapper.find('ResponseBody').length).toBe(1);

  const doc = wrapper.find(Response);
  doc.instance().setTab('metadata');

  // I want to do the below assertion instead, but it's not working
  // expect(doc.find('ResponseMetadata').length).toBe(1);
  expect(doc.html().includes('Response Headers')).toBe(true);

  // Should include request body in HTML
  expect(doc.html().includes('Request Data')).toBe(true);
  expect(doc.html().includes(JSON.stringify({ b: 2 }))).toBe(true);
});

describe('renders ResponseMetadata correctly', () => {
  test('renders 6 Meta if requestBody is set', (done) => {
    const wrapper = mount(
      <IntlProvider>
        <Response
          {...props}
          result={{
            status: 200,
            responseBody: JSON.stringify({ a: 1 }),
            requestBody: JSON.stringify({ b: 2 }),
            requestHeaders: [],
            method: 'post',
            responseHeaders: [],
          }}
        />
      </IntlProvider>
    );
    const doc = wrapper.find(Response);
    doc.setState({
      responseTab: 'metadata'
    })
    wrapper.update()
    setTimeout(() => {
      const metadata = wrapper.find(ResponseMetadata)
      expect(metadata.find('Meta')).toHaveLength(6)
      done()
    })
  })

  test('renders 5 Meta if requestBody is not set', (done) => {
    const wrapper = mount(
      <IntlProvider>
        <Response
          {...props}
          result={{
            status: 200,
            responseBody: JSON.stringify({ a: 1 }),
            requestBody: null,
            requestHeaders: [],
            method: 'post',
            responseHeaders: [],
          }}
        />
      </IntlProvider>
    );
    const doc = wrapper.find(Response);
    doc.setState({
      responseTab: 'metadata'
    })
    wrapper.update()
    setTimeout(() => {
      const metadata = wrapper.find(ResponseMetadata)
      expect(metadata.find('Meta')).toHaveLength(5)
      done()
    })
  })

  test('renders Response headers correctly', (done) => {
    const resultResponseHeaders = [
      {
        name: 'header-1',
        value: 'value-header-1'
      }, {
        name: 'header-2',
        value: 'value-header-2'
      }, {
        name: 'header-3',
        value: 'value-header-3'
      }, {
        name: 'header-no-value',
        value: null
      }
    ]
    const wrapper = mount(
      <IntlProvider>
        <Response
          {...props}
          result={{
            status: 200,
            responseBody: JSON.stringify({ a: 1 }),
            requestBody: null,
            requestHeaders: [],
            method: 'post',
            responseHeaders: resultResponseHeaders,
          }}
        />
      </IntlProvider>
    );
    const doc = wrapper.find(Response);
    doc.setState({
      responseTab: 'metadata'
    })
    wrapper.update()
    setTimeout(() => {
      const metadata = wrapper.find(ResponseMetadata)
      const responseHeaders = metadata.findWhere(node => node.prop('label') === 'Response Headers')
      expect(responseHeaders).toHaveLength(1)
      expect(responseHeaders.find('span')).toHaveLength(resultResponseHeaders.length*2)
      expect(responseHeaders.html()).toMatchSnapshot()
      done()
    })
  })
})
