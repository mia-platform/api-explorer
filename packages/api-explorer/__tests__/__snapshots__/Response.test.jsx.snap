// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ResponseMetadata correctly renders Response headers correctly 1`] = `"<div style=\\"display: grid; grid-template-columns: 1fr; grid-template-rows: auto 1fr; padding: 5px;\\"><label style=\\"text-transform: uppercase; font-size: 11px; margin-bottom: 3px; color: rgba(255, 255, 255, 0.4);\\">Response Headers</label><div style=\\"margin: 0px 10px; padding: 0px; font-size: 11px; font-family: monospace; color: rgb(255, 255, 255);\\"><div style=\\"display: grid;\\"><div style=\\"display: flex; justify-content: space-between; padding: 3px 5px;\\"><span style=\\"font-style: italic; padding-right: 5px; font-size: 10px;\\">header-1</span><span style=\\"font-weight: bold; font-size: 12px; word-break: break-all;\\">value-header-1</span></div><div style=\\"display: flex; justify-content: space-between; padding: 3px 5px;\\"><span style=\\"font-style: italic; padding-right: 5px; font-size: 10px;\\">header-2</span><span style=\\"font-weight: bold; font-size: 12px; word-break: break-all;\\">value-header-2</span></div><div style=\\"display: flex; justify-content: space-between; padding: 3px 5px;\\"><span style=\\"font-style: italic; padding-right: 5px; font-size: 10px;\\">header-3</span><span style=\\"font-weight: bold; font-size: 12px; word-break: break-all;\\">value-header-3</span></div><div style=\\"display: flex; justify-content: space-between; padding: 3px 5px;\\"><span style=\\"font-style: italic; padding-right: 5px; font-size: 10px;\\">header-no-value</span><span style=\\"font-weight: bold; font-size: 12px; word-break: break-all;\\"></span></div></div></div></div>"`;
